### DISCUSSION TITLE
eglot only shows documentation for the function call under the cursor

### DISCUSSION THREAD

**Discussion by bit69tream** [joaotavora/eglot#1173] [General]
I am using eglot that came with my emacs build (`GNU Emacs 29.0.60 (build 1, x86_64-pc-linux-gnu, GTK+ Version 3.24.35, cairo version 1.17.6) of 2023-02-03`).
I believe it should contain the last stable version of eglot.

My issue is that when I try to lookup documentation for a variable which is used as a function argument I only get documentation for that function. Doesn't matter if I just hover over it with my cursor or press the `eglot-doc-buffer` keybind.
![20-02-23--22-25-28](https://user-images.githubusercontent.com/43259894/220188277-a39fda72-c298-45f4-8938-93fcb314cd4e.png)

So if I ever want to look at the variable(or even like an enum value, etc), I must copy its name and paste it somewhere else before I can see its docs.

Like this:
![20-02-23--22-50-26](https://user-images.githubusercontent.com/43259894/220188395-ec8376b2-6068-4aaf-8377-fdb4361a2000.png)

And I checked if `lsp-mode` can achieve what I want, and it can. So its not impossible.

**Comment by joaotavora** [comment]
This is a known problem that I've been trying to solve, but it's not incredibly easy (although it looks easy)

For now, you can do this workaround.

```lisp
(add-hook 'eglot-managed-mode-hook 
          (lambda () (setq eldoc-documentation-strategy 
                           #'eldoc-documentation-compose))
```

It should do what you want.

It's possible or even likely that, if you're into having all possible information at point be available through ElDoc, you *also* want

```lisp
(add-hook 'eglot-managed-mode-hook
          (lambda () (setq eldoc-documentation-functions
                           '(flymake-eldoc-function
                             eglot-signature-eldoc-function
                             eglot-hover-eldoc-function))))
```

Then if you happen to have some syntax error or diagnostic at point, that information will also be displayed in Eldoc outlets.

The thing is, I want you and many others to not need this workaround anymore. This `eldoc-documentation-compose` should be the default anyway, as it is the one which naturally provides most information.  And then Eglot wouldn't need to mess with it, which is the right thing.

But if we make it the default, it has subtle implications in `emacs-lisp-mode`, unless we also change the default of `eldoc-echo-area-use-multiline-p` in Emacs Lisp mode.  

This will take some discussion in the Emacs bug tracker, probably.  But I think I can do part of this change soon anyway.

**Comment by joaotavora** [comment]
Related #894, #920, #648 and possibly some others.

**Comment by joaotavora** [comment]
Yep, also https://github.com/joaotavora/eglot/issues/1031

**Comment by 6b6279** [comment]
Does `eglot` set `eldoc-documentation-strategy` at all? Is the work-around with the hook necessary, even when the user explicitly calls `(setq eldoc-documentation-strategy #'eldoc-documentation-compose)` elsewhere in their init file?

**Comment by joaotavora** [comment]
> Is the work-around with the hook necessary, even when the user explicitly calls (setq eldoc-documentation-strategy #'eldoc-documentation-compose) elsewhere in their init file?

Yes it is, because eglot sets it buffer-locally.  Go ahead and test. But I'm about to change that, so that that very line in the init file will be enough.

Then I will argue in emacs-devel or bug-gnu-emacs that the global default for `eldoc-documentation-strategy` should be `eldoc-documentation-compose` to start with.

**Comment by joaotavora** [comment]
But that change, if it happens, is probably not going to be so sudden.

**Comment by 6b6279** [comment]
I think `eglot` respecting the user’s choice of `eldoc-documentation-strategy` is definitely a good idea. Thanks a lot for the change!

**Comment by bit69tream** [comment]
@joaotavora Thanks for the solution, everything works now!

**Comment by joaotavora** [comment]
Thanks. By the way, this should have been a discussion, not an issue report.  Moving to discussion.
